examples/gno.land/r/gnoland/valopers/valopers.gno

```diff
@@ -0,0 +1,468 @@
+// Package valopers is designed around the permissionless lifecycle of valoper profiles.
+package valopers
+
+import (
+	"crypto/bech32"
+	"errors"
+	"regexp"
+	"std"
+	"strconv"
+	"strings"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/avl/pager"
+	"gno.land/p/demo/combinederr"
+	"gno.land/p/demo/ufmt"
+)
+
+var (
+	ErrValoperExists       = errors.New("valoper already exists")
+	ErrValoperMissing      = errors.New("valoper does not exist")
+	ErrAddressExists       = errors.New("valoper updated address exists")
+	ErrSameAddress         = errors.New("valoper updated address is the same")
+	ErrInvalidAddress      = errors.New("invalid address")
+	ErrOriginCaller        = errors.New("origin call is not the same")
+	ErrInvalidMoniker      = errors.New("moniker cannot be empty")
+	ErrInvalidDescription  = errors.New("description cannot be empty")
+	ErrInvalidPubKey       = errors.New("invalid public key")
+	ErrInvalidP2PAddress   = errors.New("invalid P2P address")
+	ErrP2PAddressesMissing = errors.New("P2P addresses cannot be empty")
+)
+
+var (
+	valopers     *avl.Tree                  // valopers keeps track of all the active validator operators. Address -> Valoper
+	instructions string                     // markdown instructions for valoper's registration
+	minFee       int64     = 20 * 1_000_000 // minimum gnot must be paid to register.
+	minCoin                = std.NewCoin("ugnot", minFee)
+)
+
+// Valoper represents a validator operator profile
+type Valoper struct {
+	Moniker     string // A human-readable name
+	Description string // A description and details about the valoper
+
+	Address      std.Address // The bech32 gno address of the validator
+	PubKey       string      // The bech32 public key of the validator
+	Active       bool        // Flag indicating if the valoper is active
+	P2PAddresses []string    // The publicly reachable P2P addresses of the validator
+
+	Caller std.Address // The address of the caller
+}
+
+// Register registers a new valoper
+func Register(moniker string, description string, address std.Address, pubKey string, p2pAddresses string) {
+	// Check if the valoper is already registered
+	if isValoper(address) {
+		panic(ErrValoperExists)
+	}
+
+	if err := validateSentCoins(); err != nil {
+		panic(err)
+	}
+
+	v := Valoper{
+		Moniker:      moniker,
+		Description:  description,
+		Address:      address,
+		PubKey:       pubKey,
+		Active:       true,
+		P2PAddresses: splitP2PAddresses(p2pAddresses),
+		Caller:       std.PreviousRealm().Address(),
+	}
+
+	if err := v.Validate(); err != nil {
+		panic(err)
+	}
+
+	// TODO add address derivation from public key
+	// (when the laws of gno make it possible)
+
+	// Save the valoper to the set
+	valopers.Set(v.Address.String(), v)
```

Comment:
i suggest switching to a list, like with p/moul/ulist (or avl + seqid) so that we can expect changing the address later, for key migration. cc @ilgooz

Reply:
actually that's fine to do it later. up to you.

Reply:
Currently, in the valopers avl.Tree, the key is the validator address and the value is the Valoper struct. The address as the key permits us to check against duplicate valopers without parsing the whole avl.Tree (few gas consuming).
If I understand correctly, you want to use an avl.Tree + p/moul/ulist.If so we can track a key rotation? If yes, what would be the avl.Tree key? We cannot use the address because it could change. Maybe the moniker?

Reply:
Allowing valopers to change their address is a major change of how we should identify the valoper. As you said, let's do that later. For now, if a valoper needs to update its address, it can make a new registration.